photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala

```diff
@@ -0,0 +1,250 @@
+   * @param multiplyVector The gradient direction vector
+   * @param normalizationContext The normalization context
+   * @return The Hessian diagonal (multiplied by the gradient direction) of the objective function and regularization
+   *         terms
+   */
+  abstract override protected[ml] def hessianVector(
+      input: Data,
+      coefficients: Coefficients,
+      multiplyVector: Coefficients,
+      normalizationContext: BroadcastWrapper[NormalizationContext]): Vector[Double] =
+    super.hessianVector(input, coefficients, multiplyVector, normalizationContext) +
+      l2RegHessianVector(convertToVector(multiplyVector))
+
+  /**
+   * Compute the Hessian diagonal of the objective function over the given data for the given model coefficients, with
+   * regularization towards the prior coefficients.
+   *
+   * @param input The data over which to compute the Hessian diagonal
+   * @param coefficients The model coefficients for which to compute the objective function's Hessian diagonal
+   * @return The Hessian diagonal of the objective function and regularization terms
+   */
+  abstract override protected[ml] def hessianDiagonal(input: Data, coefficients: Coefficients): Vector[Double] =
+    super.hessianDiagonal(input, coefficients) :+ l2RegHessianDiagonal
+
+  /**
+   * Compute the Hessian matrix of the objective function over the given data for the given model coefficients, with
+   * regularization towards the prior coefficients.
+   *
+   * @param input The data over which to compute the Hessian matrix
+   * @param coefficients The model coefficients for which to compute the objective function's Hessian matrix
+   * @return The Hessian matrix of the objective function and regularization terms
+   */
+  abstract override protected[ml] def hessianMatrix(input: Data, coefficients: Coefficients): DenseMatrix[Double] =
+    super.hessianMatrix(input, coefficients) + l2RegHessianMatrix
+
+  /**
+   * Compute the Hessian diagonal * gradient direction of the Gaussian regularization term for the given model
+   * coefficients.
+   *
+   * @param multiplyVector The gradient direction vector
+   * @return The Hessian diagonal of the Gaussian regularization term, with gradient direction vector
+   */
+  protected def l2RegHessianVector(multiplyVector: Vector[Double]): Vector[Double] =
+    l2RegWeight * (multiplyVector /:/ priorVariances)
+
+  /**
+   * Compute the Hessian diagonal of the Gaussian regularization term for the given model coefficients.
+   *
+   * @return The Hessian diagonal of the Gaussian regularization term
+   */
+  protected def l2RegHessianDiagonal: Vector[Double] = l2RegWeight * inversePriorVariances
+
+  /**
+   * Compute the Hessian matrix of the Gaussian regularization term for the given model coefficients.
+   *
+   * @return The Hessian matrix of the Gaussian regularization term
+   */
+  protected def l2RegHessianMatrix: DenseMatrix[Double] = l2RegWeight * diag(inversePriorVariances)
```

Comment:
Thank you for catching some mistakes in my implementation of the mathematics.

Reply:
fixed